machinelearning/pytorch.md

```diff
@@ -0,0 +1,234 @@
+dnnl_verbose,exec,cpu,convolution,gemm:ref,forward_training,src_f32::blocked:abcd:f0 wei_f32::blocked:abcde:f0 bia_f32::blocked:a:f0 dst_f32::blocked:abcd:f0,post_ops:'eltwise_bounded_relu:6;';,alg:convolution_direct,mb1_g64ic64oc64_ih112oh56kh3sh2dh0ph0_iw112ow56kw3sw2dw0pw0
+```
+If the tensor ops and shapes are still not executed with ACL gemm kernels, please raise an ssue on [ACL github](https://github.com/ARM-software/ComputeLibrary) with the operator and shape details.
+
+2. Once the tensor ops are executed with ACL gemm kernels, enable fast math mode, 'export DNNL_DEFAULT_FPMATH_MODE=BF16', to pick bfloat16 hybrid gemm kernels.
+
+3. Make sure transparent huge pages are enabled
+```
+echo always > /sys/kernel/mm/transparent_hugepage/enabled
+cat /sys/kernel/mm/transparent_hugepage/enabled
+[always] madvise never
+```
+
+4. The above triaging steps cover typical issues due to the missing runtime configurations. If you are stuck with any of these steps or if the performance is still not meeting the target, please raise an issue on [aws-graviton-getting-started](https://github.com/aws/aws-graviton-getting-started) github.
+
+# Building PyTorch from source
+
+While the packages for docker container and wheels provide stable baseline for ML application development and production, they lack the latest fixes and optimizations from the master branch. The scripts for building the torch, torch-xla and torchvision wheels are avaiable in torch-xla repo. This section provides instructions for using those scripts.
+```
+# The PyTorch and torch-xla build scripts are available in torch-xla repo
+git clone https://github.com/pytorch/xla.git
+cd xla
+
+# To build and install PyTorch 1.12 release version
+./scripts/build_torch_wheels.sh 3.8 v1.12.0
+
+# To build and install PyTorch master branch
+./scripts/build_torch_wheels.sh 3.8 nightly
+
+```
+
+# How to enable torch-xla for model inferece
+PyTorch/XLA creates a TensorFlow local server everytime a new PyTorch/XLA program is run. The XRT (XLA Runtime) client connects to the local TensorFlow server. So, to enable XLA device, set the XRT device map and worker as beolow
+```
+export XRT_DEVICE_MAP="CPU:0;/job:localservice/replica:0/task:0/device:XLA_CPU:0"
+export XRT_WORKERS="localservice:0;grpc://localhost:9002"
+```
+The below snippet highlights how easy it is to switch any  model to run on XLA. The model definition, dataloader, optimizer and training loop can work on any device. The only XLA-specific code is a couple lines that acquire the XLA device and mark the step. Calling xm.mark_step() at the end of each inference iteration causes XLA to execute its current graph and update the model’s parameters
+```python
+import torch_xla
+import torch_xla.core.xla_model as xm
+
+# instead of the standard 'cuda' or 'cpu' device, create the xla device.
+device = xm.xla_device()
+.
+.
+.
+xm.mark_step() # at the end of each inference iteration
+```
+
+**Troubleshooting xla performance issues**
```

Comment:
i'm not familar enough with XLA to know if the ppl using it will see this section makes sense.

Reply:
I think you are right. may be we can remove this section.